src/asciifile.rs

```diff
@@ -0,0 +1,277 @@
+    pub fn new(mapping: Mmap) -> Result<AsciiFile, EncodingError> {
+        if let Some(position) = mapping.iter().position(|c| !c.is_ascii()) {
+            return Err(EncodingError::NotAscii {
+                position,
+                character: mapping[position].into(),
+            });
+        }
+
+        Ok(AsciiFile { mapping })
+    }
+
+    pub fn iter(&self) -> PositionedChars<std::str::Chars<'_>> {
+        let s: &str = self;
+        PositionedChars {
+            curpos: Position { row: 0, col: 0 },
+            af: s.chars(),
+            peeked: String::new(),
+        }
+    }
+}
+
+use std::ops::Deref;
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct Position {
+    row: usize,
+    col: usize,
+}
+
+use std::fmt::{self, Display};
+
+impl Display for Position {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "{}:{}", self.row, self.col)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PositionedChar(pub Position, pub char);
+
+pub struct PositionedChars<I>
+where
+    I: Iterator<Item = char>,
+{
+    curpos: Position,
+    af: I, // form ascii file
+    peeked: String,
+}
+
+impl<I> Iterator for PositionedChars<I>
+where
+    I: Iterator<Item = char>,
+{
+    type Item = PositionedChar;
+    fn next(&mut self) -> Option<PositionedChar> {
+        let c = self.peek()?;
+        self.peeked.remove(0);
+
+        let retpos = self.curpos;
+        self.curpos.col += 1;
+        if c == '\n' {
+            self.curpos.row += 1;
+            self.curpos.col = 0;
+        }
+
+        Some(PositionedChar(retpos, c))
+    }
+}
+
+impl<I> PositionedChars<I>
+where
+    I: Iterator<Item = char>,
+{
+    pub fn try_peek_multiple(&mut self, n: usize) -> Option<&str> {
+        let peeked = self.peek_multiple(n);
+        if peeked.len() < n {
+            None
+        } else {
+            Some(peeked)
+        }
+    }
+
+    /// Peek the next n characters, without advancing the iteratior (it is
+    /// actually advanced, of course, but this is hidden by a buffer).
+    /// If there are less than n charcters left, the returned slice is shorter
+    /// than that
+    pub fn peek_multiple(&mut self, n: usize) -> &str {
+        // We already have .len() characters "in stock", so we get the remaining n-len,
+        // if there are then many
+        for _ in self.peeked.len()..n {
+            match self.af.next() {
+                Some(next) => self.peeked.push(next),
+                None => break,
+            }
+        }
+
+        &self.peeked[0..self.peeked.len().min(n)]
+    }
+
+    /// Can't use peekable, pecause we don't care about position when peeking
+    pub fn peek(&mut self) -> Option<char> {
+        self.peek_multiple(1).chars().next()
+    }
+
+    pub fn eof_reached(&mut self) -> bool {
+        self.peek().is_some()
+    }
+
+    pub fn eof_position(&mut self) -> Position {
+        if self.next().is_some() {
+            panic!("Must only be called at EOF")
```

Comment:
Well, this is only called in situations where the caller is sure that we are at EOF, so i would just `.unwrap()` and write `// Unwrap is safe because we are at EOF` anyway

Reply:
It's a public function and can be called from anywhere.